src/background/background-message-distributor.ts

```diff
@@ -29,44 +30,33 @@ export class BackgroundMessageDistributor {
         message: InterpreterMessage,
         sender?: Sender,
     ): void | Promise<any> => {
-        message.tabId = this.getTabId(message, sender);
-
-        const { messageHandled: isInterpretedUsingGlobalContext, result: globalContextResult } =
-            this.globalContext.interpreter.interpret(message);
-
-        const { messageHandled: isInterpretedUsingTabContext, result: tabContextResult } =
-            this.tryInterpretUsingTabContext(message);
+        if (sender?.tab?.id) {
+            message = { tabId: sender.tab.id, ...message };
+        }
 
-        const {
-            messageHandled: isInterpretedAsBackchannelWindowMessage,
-            response: backchannelMessageResponse,
-        } = this.postMessageContentHandler.handleMessage(message);
+        const interpreterResponse = this.eventResponseFactory.mergeInterpreterResponses([
+            this.globalContext.interpreter.interpret(message),
+            this.tryInterpretUsingTabContext(message),
+        ]);
 
-        if (
-            !isInterpretedUsingGlobalContext &&
-            !isInterpretedUsingTabContext &&
-            !isInterpretedAsBackchannelWindowMessage
-        ) {
-            this.logger.log('Unable to interpret message - ', message);
+        if (interpreterResponse.messageHandled) {
+            return interpreterResponse.result;
         }
 
-        if (globalContextResult || tabContextResult || backchannelMessageResponse) {
-            return this.createResponsePromise(
-                [globalContextResult, tabContextResult],
-                backchannelMessageResponse,
-            );
-        }
-    };
+        const postMessageResponse = this.postMessageContentHandler.handleMessage(message);
 
-    private getTabId(message: InterpreterMessage, sender?: Sender): number | null {
-        if (message != null && message.tabId != null) {
-            return message.tabId;
-        } else if (sender != null && sender.tab != null && sender.tab.id != null) {
-            return sender.tab.id;
+        if (postMessageResponse.messageHandled) {
+            if (postMessageResponse.response) {
+                return Promise.resolve(postMessageResponse.response);
+            } else {
+                return; // fire and forget
+            }
         }
 
-        return null;
-    }
+        return Promise.reject(
+            new Error(`Unable to interpret message - ${JSON.stringify(message)}`),
```

Comment:
Do we want to return a promise in this case? If the message isn't handled by the background, it may still be processed by a different context.